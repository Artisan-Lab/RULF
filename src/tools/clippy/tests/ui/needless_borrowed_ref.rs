// run-rustfix

#![warn(clippy::needless_borrowed_reference)]
#![allow(unused, clippy::needless_borrow)]

fn main() {}

fn should_lint(array: [u8; 4], slice: &[u8], slice_of_refs: &[&u8], vec: Vec<u8>) {
    let mut v = Vec::<String>::new();
    let _ = v.iter_mut().filter(|&ref a| a.is_empty());

    let var = 3;
    let thingy = Some(&var);
    if let Some(&ref v) = thingy {}

    if let &[&ref a, ref b] = slice_of_refs {}

    let &[ref a, ..] = &array;
    let &[ref a, ref b, ..] = &array;

    if let &[ref a, ref b] = slice {}
    if let &[ref a, ref b] = &vec[..] {}

    if let &[ref a, ref b, ..] = slice {}
    if let &[ref a, .., ref b] = slice {}
    if let &[.., ref a, ref b] = slice {}
}

fn should_not_lint(array: [u8; 4], slice: &[u8], slice_of_refs: &[&u8], vec: Vec<u8>) {
    if let [ref a] = slice {}
    if let &[ref a, b] = slice {}
    if let &[ref a, .., b] = slice {}

    // must not be removed as variables must be bound consistently across | patterns
    if let (&[ref a], _) | ([], ref a) = (slice_of_refs, &1u8) {}

    let mut var2 = 5;
    let thingy2 = Some(&mut var2);
    if let Some(&mut ref mut v) = thingy2 {
        //          ^ should **not** be linted
        // v is borrowed as mutable.
        *v = 10;
    }
    if let Some(&mut ref v) = thingy2 {
        //          ^ should **not** be linted
        // here, v is borrowed as immutable.
        // can't do that:
        //*v = 15;
    }
}

enum Animal {
    Cat(u64),
    Dog(u64),
}

fn foo(a: &Animal, b: &Animal) {
    match (a, b) {
        // lifetime mismatch error if there is no '&ref' before `feature(nll)` stabilization in 1.63
        (&Animal::Cat(v), &ref k) | (&ref k, &Animal::Cat(v)) => (),
        //                  ^    and   ^ should **not** be linted
        (&Animal::Dog(ref a), &Animal::Dog(_)) => (), //              ^ should **not** be linted
    }
}
