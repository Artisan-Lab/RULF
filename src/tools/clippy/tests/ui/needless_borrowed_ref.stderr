error: this pattern takes a reference on something that is being dereferenced
  --> $DIR/needless_borrowed_ref.rs:10:34
   |
LL |     let _ = v.iter_mut().filter(|&ref a| a.is_empty());
   |                                  ^^^^^^
   |
   = note: `-D clippy::needless-borrowed-reference` implied by `-D warnings`
help: try removing the `&ref` part
   |
LL -     let _ = v.iter_mut().filter(|&ref a| a.is_empty());
LL +     let _ = v.iter_mut().filter(|a| a.is_empty());
   |

error: this pattern takes a reference on something that is being dereferenced
  --> $DIR/needless_borrowed_ref.rs:14:17
   |
LL |     if let Some(&ref v) = thingy {}
   |                 ^^^^^^
   |
help: try removing the `&ref` part
   |
LL -     if let Some(&ref v) = thingy {}
LL +     if let Some(v) = thingy {}
   |

error: this pattern takes a reference on something that is being dereferenced
  --> $DIR/needless_borrowed_ref.rs:16:14
   |
LL |     if let &[&ref a, ref b] = slice_of_refs {}
   |              ^^^^^^
   |
help: try removing the `&ref` part
   |
LL -     if let &[&ref a, ref b] = slice_of_refs {}
LL +     if let &[a, ref b] = slice_of_refs {}
   |

error: dereferencing a slice pattern where every element takes a reference
  --> $DIR/needless_borrowed_ref.rs:18:9
   |
LL |     let &[ref a, ..] = &array;
   |         ^^^^^^^^^^^^
   |
help: try removing the `&` and `ref` parts
   |
LL -     let &[ref a, ..] = &array;
LL +     let [a, ..] = &array;
   |

error: dereferencing a slice pattern where every element takes a reference
  --> $DIR/needless_borrowed_ref.rs:19:9
   |
LL |     let &[ref a, ref b, ..] = &array;
   |         ^^^^^^^^^^^^^^^^^^^
   |
help: try removing the `&` and `ref` parts
   |
LL -     let &[ref a, ref b, ..] = &array;
LL +     let [a, b, ..] = &array;
   |

error: dereferencing a slice pattern where every element takes a reference
  --> $DIR/needless_borrowed_ref.rs:21:12
   |
LL |     if let &[ref a, ref b] = slice {}
   |            ^^^^^^^^^^^^^^^
   |
help: try removing the `&` and `ref` parts
   |
LL -     if let &[ref a, ref b] = slice {}
LL +     if let [a, b] = slice {}
   |

error: dereferencing a slice pattern where every element takes a reference
  --> $DIR/needless_borrowed_ref.rs:22:12
   |
LL |     if let &[ref a, ref b] = &vec[..] {}
   |            ^^^^^^^^^^^^^^^
   |
help: try removing the `&` and `ref` parts
   |
LL -     if let &[ref a, ref b] = &vec[..] {}
LL +     if let [a, b] = &vec[..] {}
   |

error: dereferencing a slice pattern where every element takes a reference
  --> $DIR/needless_borrowed_ref.rs:24:12
   |
LL |     if let &[ref a, ref b, ..] = slice {}
   |            ^^^^^^^^^^^^^^^^^^^
   |
help: try removing the `&` and `ref` parts
   |
LL -     if let &[ref a, ref b, ..] = slice {}
LL +     if let [a, b, ..] = slice {}
   |

error: dereferencing a slice pattern where every element takes a reference
  --> $DIR/needless_borrowed_ref.rs:25:12
   |
LL |     if let &[ref a, .., ref b] = slice {}
   |            ^^^^^^^^^^^^^^^^^^^
   |
help: try removing the `&` and `ref` parts
   |
LL -     if let &[ref a, .., ref b] = slice {}
LL +     if let [a, .., b] = slice {}
   |

error: dereferencing a slice pattern where every element takes a reference
  --> $DIR/needless_borrowed_ref.rs:26:12
   |
LL |     if let &[.., ref a, ref b] = slice {}
   |            ^^^^^^^^^^^^^^^^^^^
   |
help: try removing the `&` and `ref` parts
   |
LL -     if let &[.., ref a, ref b] = slice {}
LL +     if let [.., a, b] = slice {}
   |

error: aborting due to 10 previous errors

